clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
+                int LocationID = -1) {
     ParamDesc newParamDesc =
         makeParamDesc(SemaRef.getASTContext(), BS, FieldTy);
-    addParam(newParamDesc, FieldTy);
+    addParam(newParamDesc, FieldTy, LocationID);
   }
 
-  void addParam(ParamDesc newParamDesc, QualType FieldTy) {
+  void addParam(ParamDesc newParamDesc, QualType FieldTy, int LocationID) {
     // Create a new ParmVarDecl based on the new info.
+    ASTContext &Ctx = SemaRef.getASTContext();
     auto *NewParam = ParmVarDecl::Create(
-        SemaRef.getASTContext(), KernelDecl, SourceLocation(), SourceLocation(),
+        Ctx, KernelDecl, SourceLocation(), SourceLocation(),
         std::get<1>(newParamDesc), std::get<0>(newParamDesc),
         std::get<2>(newParamDesc), SC_None, /*DefArg*/ nullptr);
     NewParam->setScopeInfo(0, Params.size());
     NewParam->setIsUsed();
 
     LastParamIndex = Params.size();
     Params.push_back(NewParam);
+    if (LocationID != -1)
+      NewParam->addAttr(
+          SYCLIntelBufferLocationAttr::CreateImplicit(Ctx, LocationID));
+  }
+
+  // Obtain an integer value stored in a template parameter of buffer_location
+  // property to pass it to buffer_location kernel attribute
+  int handleBufferLocationProperty(QualType FieldTy, SourceLocation Loc) {
+    const auto *AccTy =
+        cast<ClassTemplateSpecializationDecl>(FieldTy->getAsRecordDecl());
+    // TODO: when SYCL headers' part is ready - replace this 'if' with an error
+    if (AccTy->getTemplateArgs().size() < 6)
+      return -1;
+    const auto PropList = cast<TemplateArgument>(AccTy->getTemplateArgs()[5]);
+    if (PropList.getKind() != TemplateArgument::ArgKind::Type) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
+          << "accessor's 5th" << /*type*/ 1;
+      return -1;
+    }
+    QualType PropListTy = PropList.getAsType();
+    if (!Util::isPropertyListType(PropListTy)) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
+          << "accessor's 5th" << /*property_list*/ 3;
+      return -1;
+    }
+
+    int LocationID = -1;
+    const auto *PropListDecl =
+        cast<ClassTemplateSpecializationDecl>(PropListTy->getAsRecordDecl());
+    const auto TemplArg = PropListDecl->getTemplateArgs()[0];
+    if (TemplArg.getKind() != TemplateArgument::ArgKind::Pack) {
+      SemaRef.Diag(Loc, diag::err_sycl_invalid_property_template_param)
+          << "property_list" << /*parameter pack*/ 0;
+      return -1;
+    }
+    // Move through TemplateArgs list of a property list and search for
+    // buffer_location property. If found - return the stored integer value in
+    // its template parameter, if not - return -1.
+    for (TemplateArgument::pack_iterator Prop = TemplArg.pack_begin();
+         Prop != TemplArg.pack_end(); ++Prop) {
+      QualType PropTy = Prop->getAsType();
+      if (Util::isSyclBufferLocationType(PropTy)) {
+        // If we have more than 1 buffer_location properties on a single
```

Comment:
This one is untested, but can be tested with the current version of sycl.hpp, to fix